### PR TITLE
Symmetric QGEMM 

### PR DESCRIPTION
**Pull request by chenfucn**
**Description**: 

Adding code for symmetric quantized matrix multiplication.  Used in quantized convolution, achieving significant perf gain.

TODO, use this in other operators

DOT kernel perf test:

Pixel 5a:

Cartoongan | 513.539 ms | 471.786 ms
-- | -- | --
Efficient | 57.5169 ms | 56.4174 ms
Edgetpu | 14.6673 ms | 13.5959 ms

NEON kernel perf test

Pixel 3a

Cartoongan | 1423.53 ms | 1069.92 ms
-- | -- | --
Efficient | 114.086 ms | 107.968 ms
Edgetpu | 39.2632 ms | 36.9839 ms

### CODE REVIEW

File: onnxruntime/core/providers/cpu/quantization/qlinearconv.cc [hunk 1]
```diff
@@ -141,46 +141,97 @@ class QLinearConv : public OpKernel {
...
+    // Try symmetric GEMM packing
+    // Don't pack the filter buffer if the MlasConvDepthwise path is used.
+    if (group_input_channels != 1 || group_output_channels != 1) {
+      const size_t kernel_dim = group_input_channels * kernel_size;
+      packed_W_size_ = MlasSymmQgemmPackBSize(group_output_channels,
+                                              kernel_dim,
+                                              std::is_same<ActType, int8_t>::value);
+      
+      if (packed_W_size_ != 0) {
+        size_t packed_W_data_size = SafeInt<size_t>(group_count) * packed_W_size_;
```
Comment (yufenglee) [inline onnxruntime/core/providers/cpu/quantization/qlinearconv.cc:203]: >SafeInt [](http://example.com/codeflow?start=36&length=7)

it is unnecessary as group_count is already size_t

File: onnxruntime/core/providers/cpu/quantization/qlinearconv.cc [hunk 2]
```diff
@@ -141,46 +141,97 @@ class QLinearConv : public OpKernel {
...
 
-          is_symmetric_conv_ = true;
+      is_symmetric_conv_ = true;
+      is_W_packed_ = true;
+      return true;
+    }
 
-          is_W_packed_ = true;
-          return true;
+    // Try symmetric GEMM packing
```
Comment (yufenglee) [inline onnxruntime/core/providers/cpu/quantization/qlinearconv.cc:194]: consolidate this logic with the normal gemm packing?

File: onnxruntime/core/providers/cpu/quantization/qlinearconv.cc [hunk 3]
```diff
@@ -577,9 +629,13 @@ Status QLinearConv<ActType>::Compute(OpKernelContext* context) const {
     padding_data.resize(static_cast<size_t>(C), X_zero_point_value);
   }
 
-  int32_t thread_count = ComputeThreadCount(output_image_size, group_output_channels, kernel_dim);
   concurrency::ThreadPool* thread_pool = context->GetOperatorThreadPool();
+#if defined(_M_ARM64) || defined(__aarch64__)
+  int32_t thread_count = (output_image_size + 3) / 4;
```
Comment (yufenglee) [inline onnxruntime/core/providers/cpu/quantization/qlinearconv.cc:634]: >4 [](http://example.com/codeflow?start=51&length=1)

Could you please add the comment to clarify the reason to choose 4 here (to align with the row count of kernel)?

And we need to change the name thread_count. It is actually task/work count. We can do it in a separate PR though.

File: onnxruntime/core/mlas/lib/qgemm_kernel_sdot.cpp [hunk 4]
```diff
@@ -757,3 +757,280 @@ const MLAS_GEMM_QUANT_DISPATCH MlasGemmS8S8DispatchSdot = {
...
+    //  |v4.b[3]..v4.b[15] v5.b[3]..v5.b[15]  v6.b[3]..v6.b[15]  v7.b[3]..v7.b[15]|
+    //
+    // So we process B data similar with MlasGemmQuantCopyPackB<MLAS_GEMM_S8S8_KERNEL_SDOT>
+    // But twice as wide and twice as deep:
+    //     16 CountN and 16 CountK
+    //
+    //      [ A0 A1 A2 A3 B0 B1 B2 B3 C0 C1 C2 C3 D0 D1 D2 D3 ]
+    //      [ E0 E1 E2 E3 F0 F1 F2 F3 G0 G1 G2 G3 H0 H1 H2 H3 ]
+    //      [ H4 H5 H6 H7 I4 I5 I6 I7 J4 J5 J6 J7 K4 K5 K6 K7 ]
+    //      [ L4 L5 L6 L7 M4 M5 M6 M7 N4 N5 N6 N7 O4 O5 O6 O7 ]
```
Comment (yufenglee) [inline onnxruntime/core/mlas/lib/qgemm_kernel_sdot.cpp:806]: should be I0 I1 I2 I3 ...

File: onnxruntime/core/mlas/lib/qgemm_kernel_sdot.cpp [hunk 5]
```diff
@@ -757,3 +757,280 @@ const MLAS_GEMM_QUANT_DISPATCH MlasGemmS8S8DispatchSdot = {
...
+            BytesRow[1] = vld1_s8(&PaddedMatrixBData[24]);
+            BytesRow[2] = vld1_s8(&PaddedMatrixBData[40]);
+            BytesRow[3] = vld1_s8(&PaddedMatrixBData[56]);
+            MlasGemmS8S8CopyPackBProcessSDot(D, BytesRow, ColumnSums1);
+            D += 32;
+        }
+
+        //
+        // Zero pad the output buffer to a multiple of PackedK if the above
+        // processed an odd number of four row bundles.
```
Comment (yufenglee) [inline onnxruntime/core/mlas/lib/qgemm_kernel_sdot.cpp:974]: >odd [](http://example.com/codeflow?start=24&length=3)

similar to the above

File: onnxruntime/core/mlas/lib/aarch64/SymQgemmS8KernelSdot.S [hunk 6]
```diff
@@ -0,0 +1,388 @@
...
+        SdotByElement    20, 5, 8,1
+        SdotByElement    21, 5, 9,1
+        SdotByElement    22, 5,10,1
+        SdotByElement    23, 5,11,1
+        SdotByElement    24, 6, 8,1
+        SdotByElement    25, 6, 9,1
+        SdotByElement    26, 6,10,1
+        SdotByElement    27, 6,11,1
+        SdotByElement    28, 7, 8,1
+        SdotByElement    29, 7, 9,1
```
Comment (yufenglee) [inline onnxruntime/core/mlas/lib/aarch64/SymQgemmS8KernelSdot.S:317]: we need to handle the over-reading issue of A.

File: onnxruntime/core/providers/cpu/quantization/qlinearconv.cc (hunk 1 above)
Comment (chenfucn) [reply onnxruntime/core/providers/cpu/quantization/qlinearconv.cc:203]: I copied this part of the code, I guess the intention was to deal with multiplication overflow.

File: onnxruntime/core/providers/cpu/quantization/qlinearconv.cc (hunk 2 above)
Comment (chenfucn) [reply onnxruntime/core/providers/cpu/quantization/qlinearconv.cc:210]: more suggestions?

File: onnxruntime/core/framework/allocator.cc [hunk 7]
```diff
@@ -43,6 +43,7 @@ bool IAllocator::CalcMemSizeForArrayWithAlignment(size_t nmemb, size_t size, siz
 void* AllocatorDefaultAlloc(size_t size) {
   const size_t alignment = MlasGetPreferredBufferAlignment();
   if (size <= 0) return nullptr;
+  size += MLAS_SYMM_QGEMM_BUF_OVERRUN;
```
Comment (yufenglee) [inline onnxruntime/core/framework/allocator.cc:46]: What if the user uses customized allocator? I think we need to handle it in the computation of buffer size of tensor.

File: onnxruntime/core/framework/allocator.cc [hunk 8]
```diff
@@ -43,6 +43,7 @@ bool IAllocator::CalcMemSizeForArrayWithAlignment(size_t nmemb, size_t size, siz
...
   void* p;
```
Comment (skottmckay) [inline onnxruntime/core/framework/allocator.cc:47]: Why do we add a constant overrun amount vs. rounding up the size to the required boundary?

Comment (chenfucn) [reply onnxruntime/core/framework/allocator.cc:47]: Rounding up is not enough.  For instance, the activation tensor is 8 by 8. So size is 64, alignment is good. But the last row only has 8 values.  The kernel needs to have a vector register to load the last row. It takes 16 byte to fill that vector register. Thus the kernel still overread by 8 bytes.

**Review by yufenglee — APPROVED** [review]
:shipit:
Let's handle the overreading more generally in a separated PR.

File: onnxruntime/core/framework/allocator.cc (hunk 8 above)
Comment (skottmckay) [reply onnxruntime/core/framework/allocator.cc:47]: Sorry - not quite following. If the last row is 8 and you need a multiple of 16, can't you determine that only 8 additional are required? Adding 15 is obviously simpler - was just wondering if there was a reason other than that to not take into account the actual number needed.
